### ISSUE TITLE
Weblogin flow returns input loginname instead of user id

### ISSUE THREAD

**Issue by tobiasKaminsky** [nextcloud/server#7593]
When using the new weblogin flow and enter "TOBI" as loginname I also get "TOBI" back instead of "tobi" which is the internal, unique username.
However if I enter an email address it does change the username to "tobi".

Am I missing something?

**Comment by rullzer** [comment]
@tobiasKaminsky you could get the UID with a request to <server>/ocs/v2.php/cloud/user

**Comment by tobiasKaminsky** [comment]
Yes, but this is a second call and for me it seems just to be wrong. If you ask for the uid you should always get the same uid and not in case of weblogin something else.

**Comment by rullzer** [comment]
You don't ask for the uid. You ask to login. It would be the same if you use the old login flow. If you login with Admin but the uid is admin. Login etc would still work. 

Anyways. still on my todo to properly look into this but will be in the new year I guess

**Comment by blizzz** [comment]
> You don't ask for the uid. You ask to login. It would be the same if you use the old login flow. If you login with Admin but the uid is admin. Login etc would still work.

Well, we always used to return the user id on successful login as this is not known to the requesting end user, but needed for any other action.

**Comment by rullzer** [comment]
@blizzz where did we do this? In the web?

**Comment by blizzz** [comment]
https://github.com/nextcloud/server/blob/master/lib/private/User/Manager.php#L190

back back in the days it was the user id only, which is still what each backend's checkPassword returns.

**Comment by rullzer** [comment]
@blizzz as promised an explanation:

So the new flow provides us with a url like: `nc://login/server:<server>&user:<user>&pass:<pass>`

The login magic is a bit weird. As you can see in https://github.com/nextcloud/server/blob/master/core/Controller/LoginController.php#L254-L264.

* If you login with a loginname we will just set that
* If you login with an e-mail your login name is set to the userid.

Now because the login name does not have to be unique. Like in the database backend everything is lowercased we have no unique id. So returning the login name is not always usefull.

For example the new dav endpoint: `<server>/remote.php/dav/files/<principal>`. If there the login name doesn't ahve to be unique then how will the clients know what to use? (new dav features will require clients to use this new endpoint).

Same for the E2E app actually since we require the CN to be set to the uid (as again the login name doesn't have to be unique).

Long story short:

1. Client wants to be able to login easily with the new flow
2. They want to get back a unique id they can use to (in combination with the token) to
  * Auth against endpoints
  * Naviage the dav endpoint
  * Set the CN for the E2E certs

**Comment by blizzz** [comment]
> Now because the login name does not have to be unique. Like in the database backend everything is lowercased we have no unique id. So returning the login name is not always usefull.

I oughtta get another local User Management out for dev purposes that is driven by random, opaque userids. The native one does some weird magic resulting in confusion and inaccurate assumptions.

> For example the new dav endpoint: <server>/remote.php/dav/files/<principal>. If there the login name doesn't ahve to be unique then how will the clients know what to use? (new dav features will require clients to use this new endpoint).

Well, the Loginname should solely be used for authentication. The email is also just another loginname (awkward implementation, but effectively not different).

One exception: the loginname *can* be used for federated IDs, since we try to translate them their. It's a hack, but necessary, since users do not know the ID (and/or it is not handy).

> 
    If you login with a loginname we will just set that
    If you login with an e-mail your login name is set to the userid.

Always the userid should result out of this process! 

Here I am not sure where the the final return value is being composed, tryLogin() only does a redirect. On the code lines following the quotes ones, always the $loginResult is used, except some cases where (I hope) the loginname is required. I think there's some code path I am unaware of?

**Comment by rullzer** [comment]
So... I have been thinking about this. We always need the loginname. But we (in some cases) also need the uid. And we can't just return the uid since that might not be human readable. And can't always be used to login.

We could add the uid to the login flow link. So for new connections we do fetch it. Otherwise a call to <server>/ocs/v2.php/cloud/user has to be made to obtain the uid.

@tobiasKaminsky would either of those help?

**Comment by tobiasKaminsky** [comment]
Well...the real fix for me would be to have only one unique "id" for clients. Login is done via web flow, so client does not have/need any info on loginname, but the returning "id" should be usable for all and every call.
If we/client want to show a human redable username/full name/whatever I am fine with doing a call.
But having the client to know that different "ids" are used for different calls can and will only lead to problems and confusions.

**Comment by rullzer** [comment]
@blizzz the flow is documented at https://github.com/nextcloud/documentation/pull/642

The issue here is that the clients have no idea what is being entered as login name. Long story short this boils down to:

1. Login is always done with a login name
2. Once logged in use the userid

The new flow should this also return the userid (along with the loginname so they can properly auth).


However there is the issue of the current login handler if you login with an e-mail address. Then the loginname gets set to the user id. Of course login in with e-mail has other disadvantages as well since you can change your e-mail and then all of a sudden your clients refuse to function.

I think we need to move this to 14 as I'm not comfortable with hacking in something so close to the RC regarding login....

**Comment by blizzz** [comment]
Yeah, there's some stuff to reorganize and streamline. Since there is no blocking issue, let's solve this properly in 14, instead of hastily last second. → moving to 14.

**Comment by rullzer** [comment]
@tobiasKaminsky this is now solved right since you fetch the userid from the proper ocs endpoint. So just returning the login name here is enough (and as it should be).

**Comment by MorrisJobke** [comment]
> @tobiasKaminsky this is now solved right since you fetch the userid from the proper ocs endpoint. So just returning the login name here is enough (and as it should be).

Closing then. Could you document it in the login flow? I started it in https://github.com/nextcloud/documentation/pull/797 and only need the OCS API endpoint ;)
